src/charm.py

```diff
@@ -52,6 +53,32 @@ def __init__(self, *args: typing.Any) -> None:
             host=f"{self.app.name}-endpoints.{self.model.name}.svc.cluster.local",
             strip_prefix=True,
         )
+        self.databases: typing.Dict[str, DatabaseRequires] = {
```

Comment:
@yanksyoon came up with an observer pattern, would that be applicable here?

Reply:
Can you point me to an accepted example of this pattern in our charms so that I use it as a basis for consideration ?

Reply:
Perhaps ask Yanks since he came up with the proposal 🙂

Reply:
I think a pattern emerges from the need to solve something and then reusing that solution so may times that a pattern is formed, so there really should be some examples :) Then @yanksyoon  - could you elaborate here and link to some code? Thanks!

Reply:
I decided to move most of the code into its own module. This should resolve this discussion